src/sentry/utils/kafka.py

```diff
@@ -33,3 +39,121 @@ def get(self, key):
 
 
 producers = ProducerManager()
+
+
+@contextmanager
+def set_termination_request_handlers(handler):
+    # hook the new handlers
+    old_sigint = os_signal.signal(os_signal.SIGINT, handler)
+    old_sigterm = os_signal.signal(os_signal.SIGTERM, handler)
+    try:
+        # run the code inside the with context ( with the hooked handler)
+        yield
+    finally:
+        # restore the old handlers when exiting the with context
+        os_signal.signal(os_signal.SIGINT, old_sigint)
+        os_signal.signal(os_signal.SIGTERM, old_sigterm)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class SimpleKafkaConsumer(object):
+    def __init__(
+        self,
+        commit_batch_size,
+        consumer_group,
+        topic_name,
+        max_fetch_time_seconds,
+        initial_offset_reset="latest",
+        consumer_configuration=None,
+    ):
+        """
+        Base class for implementing kafka consumers.
+        """
+
+        self.commit_batch_size = commit_batch_size
+        self.topic_name = topic_name
+        self.max_fetch_time_seconds = max_fetch_time_seconds
+        self.initial_offset_reset = initial_offset_reset
+        self.consumer_group = consumer_group
+
+        cluster_name = settings.KAFKA_TOPICS[topic_name]["cluster"]
+        bootstrap_servers = settings.KAFKA_CLUSTERS[cluster_name]["bootstrap.servers"]
+
+        self.consumer_configuration = {
+            "bootstrap.servers": bootstrap_servers,
+            "group.id": consumer_group,
+            "enable.auto.commit": "false",  # we commit manually
+            "enable.auto.offset.store": "true",  # we let the broker keep count of the current offset (when committing)
+            "enable.partition.eof": "false",  # stop EOF errors when we read all messages in the topic
+            "default.topic.config": {"auto.offset.reset": initial_offset_reset},
+        }
+
+        if consumer_configuration is not None:
+            for key in six.iterkeys(consumer_configuration):
+                self.consumer_configuration[key] = consumer_configuration[key]
+
+    @abc.abstractmethod
+    def process_message(self, message):
+        """
+        This function is called for each message
+        :param message: the kafka message:
+        """
+        pass
+
+    def run(self, is_shutdown_requested=lambda: False):
+        """
+        Runs the message processing loop
+        """
+        logger.debug(
+            "Staring kafka consumer for topic:{} with consumer group:{}",
+            self.topic_name,
+            self.consumer_group,
+        )
+
+        consumer = kafka.Consumer(self.consumer_configuration)
+        consumer.subscribe([self.topic_name])
+
+        # setup a flag to mark termination signals received, see below why we use an array
+        termination_signal_received = [False]
```

Comment:
Curious: any specific reason we don't use an object attribute here?

Reply:
The signal handler is kind of global, but this could definitely be an attribute.